File: linkedlist/single/js/linkedlist.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {

    countNodes(head) {
        let current = head,
            count = 0;
        while (current != null) {
            count++;
            current = current.next;
        }
        return count;
    }

    printNodes(head) {
        let current = head,
            nodes = [];
        while (current != null) {
            nodes.push(current.value);
            current = current.next;
        }
        this.print(nodes.join(' -> '));
    }

    find(head, value) {
        let current = head;
        while (current.value != value) {
            current = current.next;
        }
        return current;
    }

    deleteFirst(head) {
        let newHead = head.next;
        return newHead;
    }

    deleteLast(head) {
        let current = head;
        while (current.next.next != null) {
            current = current.next;
        }
        current.next = null;
    }

    deleteNode(head, value) {
        let p = this.predecessor(head, value);
        let d = this.find(head, value);
        p.next = d.next;
    }

    insertBottom(head, nodeToInsert) {
        let current = head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = nodeToInsert;
    }

    insertTop(head, nodeToInsert) {
        nodeToInsert.next = head;
        return nodeToInsert;
    }

    insert(from_node, toNode) {
        from_node.next = toNode;
    }

    insertAfter(head, value, nodeToInsert) {
        let s = this.find(head, value);
        nodeToInsert.next = s.next;
        this.insert(s, nodeToInsert);
    }

    insertBefore(head, value, nodeToInsert) {
        let p = this.predecessor(head, value);
        nodeToInsert.next = p.next;
        this.insert(p, nodeToInsert)
    }

    predecessor(head, value) {
        let current = head;
        while (current.next.value != value) {
            current = current.next;
        }
        return current;
    }

    print(o) {
        console.log(o);
    }
}

const ll = new LinkedList();
let head = new Node(23);
head = ll.insertTop(head, new Node(32));
head = ll.insertTop(head, new Node(42));
head = ll.insertTop(head, new Node(24));
head = ll.insertTop(head, new Node(99));
head = ll.insertTop(head, new Node(101));
head = ll.insertTop(head, new Node(45));

ll.countNodes(head);

ll.printNodes(head);

let twentythree = ll.find(head, 23);
ll.print(twentythree);

let p42 = ll.predecessor(head, 42);
ll.print(p42);

head = ll.deleteFirst(head);
ll.printNodes(head);

ll.deleteLast(head);
ll.printNodes(head);

ll.deleteNode(head, 24);
ll.printNodes(head);

ll.insertBottom(head, new Node(123));
ll.printNodes(head);

ll.insertAfter(head, 42, new Node(234));
ll.printNodes(head);

ll.insertAfter(head, 99, new Node(345));
ll.printNodes(head);

ll.insertBefore(head, 345, new Node(22));
ll.printNodes(head);

// ll.print(head.next);

File: linkedlist/single/python/LinkedList.py
class Node :
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList :

    def __init__(self):
        self.head = None

    def insert_top(self, node_to_insert : Node):
        node_to_insert.next = self.head
        self.head = node_to_insert

    def insert_bottom(self, node_to_insert : Node):
        current = self.head
        while current.next:
            current = current.next
        current.next = node_to_insert

    def insert_after(self, value : int, node_to_insert : Node): 
        after = self.find(value)
        self.insert(after, node_to_insert)

    def insert_before(self, value : int, node_to_insert : Node):
        p = self.predecessor(value)
        self.insert(p, node_to_insert)

    def insert(self, from_node : Node, to_node : Node):
        to_node.next = from_node.next
        from_node.next = to_node

    def find(self, value : int) -> Node:
        current = self.head
        while current.value != value:
            current = current.next
        return current

    def predecessor(self, value : int) -> Node:
        current = self.head
        while current.next.value != value:
            current = current.next
        return current

    def delete_first(self):
        old_head = self.head
        new_head = old_head.next
        self.head = new_head
        del old_head

    def delete_last(self):
        current = self.head
        while current.next.next:
            current = current.next
        del current.next.next
        current.next = None

    def delete_node(self, value : int):
        d = self.find(value)
        p = self.predecessor(value)
        p.next = d.next
        del d

    def print_nodes(self):
        current = self.head
        while current:
            print(f"{current.value} -> ", end=" ")
            current = current.next
        print("NULL")

    def count_nodes(self) -> int:
        current = self.head
        node_lth=0
        while current:
            node_lth += 1
            current = current.next
        return node_lth

    @classmethod
    def get_class_name(cls):
        return cls.__name__
        

if __name__=='__main__':

    ll = LinkedList()

    ll.insert_top(Node(23))
    ll.insert_top(Node(32))
    ll.insert_top(Node(24))
    ll.insert_top(Node(42))

    ll.print_nodes()
    print(ll.count_nodes())

    ll.insert_bottom(Node(191))

    print(ll.count_nodes())
    ll.print_nodes()

    search = ll.find(42)

    p = ll.predecessor(191)
    print(p.value)

    print(ll.count_nodes())    
    ll.delete_last()
    print(ll.count_nodes())    

    ll.print_nodes()
    ll.delete_first()
    ll.delete_first()
    ll.print_nodes()

    ll.insert_top(Node(101))
    ll.insert_top(Node(102))
    ll.insert_top(Node(103))
    ll.insert_top(Node(104))

    print(ll.count_nodes())
    ll.print_nodes()

    ll.delete_node(101)
    print(ll.count_nodes())
    ll.print_nodes()

    ll.insert_after(102, Node(123))
    ll.print_nodes()

    ll.insert_before(102, Node(321))
    ll.print_nodes()


